server/controller/src/main/java/ai/starwhale/mlops/domain/job/JobScheduler.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2022 Starwhale, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ai.starwhale.mlops.domain.job;
+
+import ai.starwhale.mlops.domain.job.bo.Job;
+import ai.starwhale.mlops.domain.job.cache.HotJobHolder;
+import ai.starwhale.mlops.domain.job.status.JobStatus;
+import ai.starwhale.mlops.domain.job.step.bo.Step;
+import ai.starwhale.mlops.domain.job.step.task.TaskService;
+import ai.starwhale.mlops.domain.job.step.task.WatchableTask;
+import ai.starwhale.mlops.domain.job.step.task.WatchableTaskFactory;
+import ai.starwhale.mlops.domain.job.step.task.bo.Task;
+import ai.starwhale.mlops.domain.job.step.task.schedule.TaskScheduler;
+import ai.starwhale.mlops.domain.job.step.task.status.TaskStatus;
+import ai.starwhale.mlops.domain.job.step.task.status.watchers.TaskStatusChangeWatcher;
+import ai.starwhale.mlops.domain.job.step.task.status.watchers.TaskWatcherForPersist;
+import ai.starwhale.mlops.exception.SwValidationException;
+import ai.starwhale.mlops.exception.api.StarwhaleApiException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+import javax.validation.constraints.NotNull;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Service;
+import org.springframework.util.CollectionUtils;
+
+/**
+ * load job to JobHolder as active state
+ */
+@Slf4j
+@Service
+public class JobScheduler {
+
+    final HotJobHolder jobHolder;
+
+    final WatchableTaskFactory watchableTaskFactory;
+
+    final TaskScheduler taskScheduler;
+
+    final TaskService taskService;
+
+    public JobScheduler(HotJobHolder jobHolder,
+                        WatchableTaskFactory watchableTaskFactory,
+                        TaskScheduler taskScheduler,
+                        TaskService taskService) {
+        this.jobHolder = jobHolder;
+        this.watchableTaskFactory = watchableTaskFactory;
+        this.taskScheduler = taskScheduler;
+        this.taskService = taskService;
+    }
+
+    public Job schedule(@NotNull Job job, Boolean resumePausedOrFailTasks) {
```

Comment:
I think `load` is better than the word `schedule` and keep it in `jobloader` class

Reply:
renamed to addAndSchedule